### ISSUE TITLE
--watch fails in Docker container running on ARM Mac

### ISSUE THREAD

**Issue by ChristianSiegert** [denoland/deno#14244]
When I run Deno in a Docker container on an Apple M1 Mac and add the `--watch` flag, I get a “Function not implemented” error. Without the `--watch` flag, everything works fine.

**Steps to reproduce:**
1. Install [Docker](https://www.docker.com/get-started/) on an ARM Mac
2. In Terminal.app, execute command `docker run -it --rm denoland/deno:1.20.5 run --watch https://deno.land/std@0.134.0/examples/welcome.ts` (For the purpose of this bug report I chose a URL, but the issue is the same for watching local files.)

**Expected output:**
```
Welcome to Deno!
```

**Actual output:**
```
Watcher Process started.
error: Function not implemented (os error 38)
```

The same error is returned when executing
```
docker run -it --rm denoland/deno:1.20.5 fmt --watch
docker run -it --rm denoland/deno:1.20.5 lint --watch
```

Environment:
- M1 Mac, macOS 12.3.1 (21E258)
- Docker version 20.10.14, build a224086
- deno 1.20.5 (release, x86_64-unknown-linux-gnu), v8 10.0.139.6, typescript 4.6.2

**Comment by csabaxyz** [comment]
Have the same issue as you. For now, I found this comment:
https://github.com/denoland/deno_docker/issues/100#issuecomment-762338543

Using this seems to work:
```dockerfile
FROM lukechannings/deno:v1.20.5
```

Or your example:
```
docker run -it --rm lukechannings/deno:v1.20.5 run --watch https://deno.land/std@0.134.0/examples/welcome.ts
```

**Comment by nsadeh** [comment]
I have the same issue - not too enthused about having to use a non-official Deno image. Is there a fix in official Deno?

**Comment by thekntl** [comment]
+1 @nsadeh

**Comment by CarlosAmaral** [comment]
Agreed, would be great to have an official image, until then I'll try Luke's image

**Comment by thekntl** [comment]
I am setting up a development environment with deno fresh and docker.

I faced the same "--watch FAIL" issue with the official Deno Docker image.

Luke's image solved the "exit" problem for the docker container but the hot reload which --watch stands for is not working.

Here is my Dockerfile

```
FROM lukechannings/deno

ARG GIT_REVISION
ENV DENO_DEPLOYMENT_ID=${GIT_REVISION}

WORKDIR /app

COPY . .

EXPOSE 8000

CMD ["task", "start"]
```

**Comment by ChristianSiegert** [comment]
@thekntl `lukechannings/deno` is working correctly with `--watch`. Your setup is wrong: You copy all files into the docker image which then never change. Instead, you must attach your host project directory as volume to the container.

Example:
```ts
// main.ts
console.log("hello");
```

```sh
docker run -it --rm -v $PWD:/app -w /app lukechannings/deno run --watch main.ts
```
Then edit file `main.ts` to see that the file change was detected and the process was restarted.

**Comment by nottakis** [comment]
I've faced the same "--watch FAIL" issue when I used Deno image offered by VSCode. 
Changing files like these works for me at this moment. 

Dockerfile
```
FROM lukechannings/deno:v1.30.3
```

devcontainer.json
```
{
	"name": "Deno",
 	"build": {
		"dockerfile": "Dockerfile"
	},
	
	// ...

	// Change remoteUser from vscode to deno 
	"remoteUser": "deno"
}
```

Without changing remoteUser, docker emits error and it doesn't work. 
When I use `lukechannings/deno:v1.20.5`, I was required to upgrade Deno. 

I encountered this issue while seeking a way to develop Deno with Fresh on the devContainer. Detail may be different depending on the situation.

**Comment by thekntl** [comment]
@ChristianSiegert 

Actually, I'm using Docker Compose to up the project. I used bind mount my local files and docker ```/app``` folder.

My docker-compose.yaml file:

```
...

  web:
    ...
    volumes:
      - ./:/app
```

Also container logs the following but browser content doesn't update.

```
2023-02-16 06:56:55 Watcher File change detected! Restarting!
2023-02-16 06:56:55 Listening on http://localhost:8000/
```

**Comment by ChristianSiegert** [comment]
@thekntl Turn off caching in your browser to check if that is the reason you see stale content. Also try https://github.com/denoland/deno/issues/14244#issuecomment-1430966787 to rule out that `lukechannings/deno` is the source of your problem. You can replace `console.log()` with [serve()](https://deno.land/manual@v1.30.3/examples/http_server#using-the-stdhttp-library) if you desire to test with your browser.

And to avoid misunderstandings: You mentioned hot reloading. `deno run --watch` does not reload your browser tab. You still have to reload the page manually.

**Comment by shivam-tripathi** [comment]
I am also trying to run it via docker-compose. While it prints `Watcher Process started.` - on code change it doesn't reload the server. I checked the contents of the source file within the docker container, and they reflected having new content but the watcher did not reload. :(

**Comment by und-miller** [comment]
@shivam-tripathi I can confirm the same behavior.

**Comment by und-miller** [comment]
This issue has been open for over two years. Are there any plans to resolve this issue in the near future?

**Comment by und-miller** [comment]
I recently discovered that docker-compose recently started shipping with watch functionality via this stack-overflow post: https://stackoverflow.com/a/78124025/2301459 

The documentation for docker-compose watch can be found here: https://docs.docker.com/compose/how-tos/file-watch/#use-watch

**Comment by kiratrizon** [comment]
Hello... is this still a bug?

**Comment by bartlomieju** [comment]
Closing this as no longer reproducible. The root cause was never in Deno: back in 2022, Docker Desktop on Apple Silicon emulated x86_64 Linux images via QEMU, which did not implement the `inotify_init`/`fanotify` syscalls that Deno's file watcher relies on, so those syscalls returned `ENOSYS` (`Function not implemented`, os error 38).

I re-ran the exact repro on an Apple Silicon host with current Docker Desktop:

```
docker run --rm --platform linux/amd64 denoland/deno:1.20.5 run --watch https://deno.land/std@0.134.0/examples/welcome.ts
```

It now works, even with the same old x86_64 1.20.5 binary:

```
Watcher Process started.
Check https://deno.land/std@0.134.0/examples/welcome.ts
Welcome to Deno!
Watcher Process finished. Restarting on file change...
```

Docker Desktop has since switched x86 emulation on macOS from QEMU to Rosetta, which passes inotify through correctly, so the issue is resolved at the emulation layer. If you hit a `--watch` failure on a current Deno version, please open a new issue with details.
